### PR TITLE
Fix TensorBoard error on categorical choices of mixed types

### LINKED ISSUE #4761: TensorBoard integration fails if None is suggested
### Expected behavior

`Optuna` with `TensorBoardCallback` should allow `None` as a suggestion.

### Environment

- Optuna version:3.2.0
- Python version:3.9.13
- OS:Linux-5.4.228-132.418.amzn2.x86_64-x86_64-with-glibc2.31

### Error messages, stack traces, or logs

```shell
File /opt/python3.9/site-packages/tensorboard/plugins/hparams/summary_v2.py:507, in Discrete.__init__(self, values, dtype)
    505         raise ValueError("Empty domain with no dtype specified")
    506 if dtype not in (int, float, bool, str):
--> 507     raise ValueError("Unknown dtype: %r" % (dtype,))
    508 self._dtype = dtype
    509 for value in self._values:

ValueError: Unknown dtype: <class 'NoneType'>
```
```


### Steps to reproduce

1. If I run this
```python
import optuna
from optuna.integration.tensorboard import TensorBoardCallback

from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score

X, y = make_classification(n_samples=1000, n_features=4,
                           n_informative=2, n_redundant=0,
                           random_state=0, shuffle=False)

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.33, random_state=0)

def objective(trial):
    params = {
        'max_leaf_nodes': trial.suggest_categorical('max_leaf_nodes', [None, 2, 5, 10, 20]),
    }
    clf = RandomForestClassifier()
    clf.fit(X_train, y_train)

    y_val_pred = clf.predict(X_val)
    f1 = f1_score(y_val, y_val_pred)
    return f1

tensorboard_callback = TensorBoardCallback("logs/", metric_name="f1")
study = optuna.create_study(direction="minimize")

study.optimize(objective, n_trials=10, timeout=600, callbacks=[tensorboard_callback])
```

It would return an error above.

2. If I don't have `None` in the suggestion or not use `TensorBoardCallback` then it would work fine.

### Additional context (optional)

_No response_

### PR DESCRIPTION
**Pull request by ciffelia**
## Motivation

The current version of TensorBoard integration fails when a choice of categorical distribution is not type of `int`, `float`, `bool`, or `str`. This is caused by a limitation of TensorBoard.

https://github.com/tensorflow/tensorboard/blob/80ccab8c8729ff7f8f5e17c132af808b0efe5f5b/tensorboard/plugins/hparams/summary_v2.py#L477-L499

## Description of the changes

This PR fixes the issue by converting choices to strings when needed.

Fixes #4761.

### CODE REVIEW

File: tests/integration_tests/test_tensorboard.py [hunk 1]
```diff
@@ -63,6 +63,36 @@ def objective(trial: optuna.trial.Trial) -> float:
...
+
+    tbcallback = TensorBoardCallback(dirname, metric_name)
+    study = optuna.create_study(study_name=study_name)
+    study.optimize(objective, n_trials=1, callbacks=[tbcallback])
+
+
+def test_categorical_mixed_types() -> None:
+    def objective(trial: optuna.trial.Trial) -> float:
+        x = trial.suggest_categorical("x", [None, 1, 2, 3, True, "foo"])
+        assert isinstance(x, str)
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_tensorboard.py:84]: I'm not sure why we can ensure that `x` is str.

File: optuna/integration/tensorboard.py [hunk 2]
```diff
@@ -74,9 +81,24 @@ def _add_distributions(
...
+                        "Choices contains mixed types, which is not supported by TensorBoard."
+                        "Converting all choices to strings."
+                    )
+                    choices = tuple(map(str, choices))
+                elif dtype not in (int, float, bool, str):
+                    _logger.warning(
+                        f"Choices are of type {dtype}, which is not supported by TensorBoard."
+                        "Converting all choices to strings."
+                    )
+                    choices = tuple(map(str, choices))
```
Comment (toshihikoyanase) [inline optuna/integration/tensorboard.py:97]: Maybe I missed something, but after this conversion, all the element of `choices` will be str. So, I guess `dtype` should be updated to `str` as well.

File: tests/integration_tests/test_tensorboard.py [hunk 3]
```diff
@@ -63,6 +63,36 @@ def objective(trial: optuna.trial.Trial) -> float:
...
+
+    tbcallback = TensorBoardCallback(dirname, metric_name)
+    study = optuna.create_study(study_name=study_name)
+    study.optimize(objective, n_trials=1, callbacks=[tbcallback])
+
+
+def test_categorical_mixed_types() -> None:
+    def objective(trial: optuna.trial.Trial) -> float:
+        x = trial.suggest_categorical("x", [None, 1, 2, 3, True, "foo"])
+        assert type(x) in (int, float, bool, str)
```
Comment (not522) [inline tests/integration_tests/test_tensorboard.py:84]:
```suggestion
        assert x is None or isinstance(x, (int, float, bool, str))
```

File: tests/integration_tests/test_tensorboard.py [hunk 4]
```diff
@@ -63,6 +63,36 @@ def objective(trial: optuna.trial.Trial) -> float:
...
+        return len(str(x))
+
+    dirname = tempfile.mkdtemp()
+    metric_name = "target"
+    study_name = "test_tensorboard_integration"
+
+    tbcallback = TensorBoardCallback(dirname, metric_name)
+    study = optuna.create_study(study_name=study_name)
+    study.optimize(objective, n_trials=1, callbacks=[tbcallback])
```
Comment (not522) [inline tests/integration_tests/test_tensorboard.py:93]:
```suggestion
    study.optimize(objective, n_trials=10, callbacks=[tbcallback])
```

**Review by not522 — COMMENTED** [review]
Thank you for your PR! It's almost LGTM.
I added comments for the test. Could you check them?

File: optuna/integration/tensorboard.py [hunk 5]
```diff
@@ -74,9 +81,24 @@ def _add_distributions(
...
+                        "Choices contains mixed types, which is not supported by TensorBoard. "
+                        "Converting all choices to strings."
+                    )
+                    choices = tuple(map(str, choices))
+                elif dtype not in (int, float, bool, str):
+                    _logger.warning(
+                        f"Choices are of type {dtype}, which is not supported by TensorBoard. "
+                        "Converting all choices to strings."
+                    )
+                    choices = tuple(map(str, choices))
```
Comment (toshihikoyanase) [inline optuna/integration/tensorboard.py:97]: Codecov suggested that this type conversion is not covered by unit tests.
@not522 Do you think we need a test case?

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. Looks solid to me.

**Review by not522 — APPROVED** [review]
Thanks for the update! There was a minor mypy error that I fixed. LGTM!

**Pull request merged**